### PR TITLE
[SPARK-10323] [SQL] fix nullability of In/InSet/ArrayContain

### PR DESCRIPTION
**Pull request by davies**
After this PR, In/InSet/ArrayContain will return null if value is null, instead of false. They also will return null even if there is a null in the set/array.

### CODE REVIEW

**Comment by davies** [comment]
cc @yhuai

**Comment by yhuai** [comment]
Here is the output of some sample tests using hive 1.2.1
```
hive> select cast(null as string) in ('1', cast(null as string));
OK
NULL
Time taken: 0.042 seconds, Fetched: 1 row(s)
hive> select array_contains(array('1', cast(null as string)), cast(null as string));
OK
false
Time taken: 0.042 seconds, Fetched: 1 row(s)
hive> 
```

**Comment by yhuai** [comment]
postgresql's output regarding `in`...

```
yhuai=# select cast(null as char(10)) in ('1', cast(null as char(10)));          ?column? 
----------
 
(1 row)

yhuai=# select cast(null as char(10)) in ('1', cast(null as char(10))) is null;
 ?column? 
----------
 t
(1 row)
```

**Comment by yhuai** [comment]
OK. I guess the main question at here is if we want to have a different semantic with hive on `array_contains` regarding `null`.

**Comment by davies** [comment]
From PostgresSQL:
```
If the array expression yields a null array, the result of ANY will be null. If the left-hand expression yields null, the result of ANY is ordinarily null (though a non-strict comparison operator could possibly yield a different result). Also, if the right-hand array contains any null elements and no true comparison result is obtained, the result of ANY will be null, not false (again, assuming a strict comparison operator). This is in accordance with SQL's normal rules for Boolean combinations of null values.
```

It's more consistent in PostgresSQL, I'd like to follow it. 

cc @rxin @marmbrus

**Comment by rxin** [comment]
I'd follow postgres here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -145,46 +143,43 @@ case class ArrayContains(left: Expression, right: Expression)
     }
   }
 
-  override def nullable: Boolean = false
+  override def nullable: Boolean = {
+    val dt = right.dataType
+    left.nullable || dt.isInstanceOf[ArrayType] && dt.asInstanceOf[ArrayType].containsNull
+  }
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Looks like we need to swap `left` and `right`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -145,46 +143,42 @@ case class ArrayContains(left: Expression, right: Expression)
...
-      if (!${arrGen.isNull}) {
-        ${elementGen.code}
-        if (!${elementGen.isNull}) {
-          for (int $i = 0; $i < ${arrGen.primitive}.numElements(); $i ++) {
-            if (${ctx.genEqual(right.dataType, elementGen.primitive, getValue)}) {
-              ${ev.primitive} = true;
-              break;
-            }
-          }
+    nullSafeCodeGen(ctx, ev, (arr, value) => {
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Do we need to call {defineCodeGen}?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantFoldingSuite.scala [hunk 3]
```diff
@@ -269,15 +269,5 @@ class ConstantFoldingSuite extends PlanTest {
       testRelation
         .select('a)
         .where(In(Literal(1), Seq(Literal(1), 'a.attr)))
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantFoldingSuite.scala]: Remove this?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 4]
```diff
@@ -382,15 +378,13 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       df.selectExpr("array_contains(null, 1)")
     }
 
-    // In hive, if either argument has a matching type has a null value, return false, even if
-    // the first argument array contains a null and the second argument is null
     checkAnswer(
-      df.selectExpr("array_contains(array(array(1), null)[1], 1)"),
-      Seq(Row(false), Row(false))
+      df.selectExpr("array_contains(array(array(1), null)[0], 1)"),
+      Seq(Row(true), Row(true))
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Do you want to test something like `array_contains(null, 1)`?

**Comment by yhuai** [comment]
+1 to follow postgres.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nullSafeCodeGen is similar to defineCodeGen

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala (hunk 4 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala]: Will add this to CollectionFunctionSuite

**Comment by yhuai** [comment]
LGTM

**Comment by davies** [comment]
Merged into master and 1.5 branch.

**Pull request closed without merging**